### PR TITLE
stm32f0-gpio-update

### PR DESCRIPTION
**Pull request by MohmadAyman**
updating the stm32f0 driver the same way as PR https://github.com/RIOT-OS/RIOT/pull/3535 
issue #2726 ,

### CODE REVIEW

File: cpu/stm32f0/vectors.c [hunk 1]
```diff
@@ -95,9 +93,9 @@ ISR_VECTORS const void *interrupt_vector[] = {
     (void*) isr_rtc,                /* real time clock */
     (void*) isr_flash,              /* flash memory controller */
     (void*) isr_rcc,                /* reset and clock control */
-    (void*) isr_exti0_1,            /* external interrupt lines 0 and 1 */
-    (void*) isr_exti2_3,            /* external interrupt lines 2 and 3 */
-    (void*) isr_exti4_15,           /* external interrupt lines 4 to 15 */
+    (void*) isr_exti,            /* external interrupt lines 0 and 1 */
+    (void*) isr_exti,            /* external interrupt lines 0 and 1 */
```
Comment (haukepetersen) [inline cpu/stm32f0/vectors.c]: lines 2 and 3 :-)

File: cpu/stm32f0/vectors.c [hunk 2]
```diff
@@ -95,9 +93,9 @@ ISR_VECTORS const void *interrupt_vector[] = {
...
+    (void*) isr_exti,            /* external interrupt lines 0 and 1 */
```
Comment (haukepetersen) [inline cpu/stm32f0/vectors.c]: 4 to 15...

File: cpu/stm32f0/periph/gpio.c [hunk 3]
```diff
@@ -341,209 +152,77 @@ int gpio_init_int(gpio_t dev, gpio_pp_t pullup, gpio_flank_t flank, gpio_cb_t cb
...
+        gpio_set(pin);
     }
 }
 
-void gpio_write(gpio_t dev, int value)
+void gpio_write(gpio_t pin, int value)
 {
     if (value) {
-        gpio_set(dev);
+        gpio_set(pin);
```
Comment (haukepetersen) [inline cpu/stm32f0/periph/gpio.c]: you could gain slightly more efficiency by saving the function call here: 

s/gpio_set(pin);/_port(pin)->BSRRL = (1 << _pin_num(pin));/

Comment (MohmadAyman) [inline cpu/stm32f0/periph/gpio.c]: you are right, i'll change it

File: cpu/stm32f0/vectors.c (hunk 2 above)
Comment (MohmadAyman) [inline cpu/stm32f0/vectors.c]: Thanks for the hints

File: cpu/stm32f0/periph/gpio.c [hunk 4]
```diff
@@ -24,315 +24,126 @@
...
     }
-    NVIC_EnableIRQ(gpio_irq_map[dev]);
 
-    /* set callback */
-    gpio_config[dev].cb = cb;
-    gpio_config[dev].arg = arg;
+
 
     /* configure the event that triggers an interrupt */
     switch (flank) {
```
Comment (haukepetersen) [inline cpu/stm32f0/periph/gpio.c]: This you could also optimize slightly.
You override the value of the `gpio_flank_t` type in the `periph_cpu.h` to these values:
```c
#define HAVE_GPIO_FLANK_T
typedef enum {
    GPIO_FALLING = 1,       /**< emit interrupt on falling flank */
    GPIO_RISING = 2,        /**< emit interrupt on rising flank */
    GPIO_BOTH = 3           /**< emit interrupt on both flanks */
} gpio_flank_t;
```

then drop the switch statement and write this:
```c
EXTI->RTSR &= ~(1 << pin_num);
EXTI->RTSR |= (((flank >> 1) & 0x1) << pin_num);
EXTI->FTSR &= ~(1 << pin_num);
EXTI->FTSR |= ((flank & 0x1) << pin_num);
```

File: cpu/stm32f0/periph/gpio.c [hunk 5]
```diff
@@ -24,315 +24,126 @@
...
-            GPIO_11_EXTI_CFG();
-            break;
-#endif
+
+    /* initialize pin as input */
+    gpio_init(pin, GPIO_DIR_IN, pullup);
+
+    /* enable global pin interrupt */
+    if (pin_num < 5) {
+        NVIC_EnableIRQ(EXTI0_IRQn + pin_num);
```
Comment (haukepetersen) [inline cpu/stm32f0/periph/gpio.c]: these seem to be off: this should be
```c
if (pin_num < 2) {
    NVIC_EnableIRQ(EXTI0_1_IRQn + pin_num);
}
else if (pin_num < 4) {
    NVIC_EnableIRQ(EXTI2_3_IRQn + pin_num);
...
```

File: cpu/stm32f0/periph/gpio.c [hunk 6]
```diff
@@ -24,315 +24,126 @@
...
+    /* set callback */
+    exti_ctx[pin_num].cb = cb;
+    exti_ctx[pin_num].arg = arg;
 
-    /* set interrupt priority (its the same for all EXTI interrupts) */
-    NVIC_SetPriority(EXTI0_1_IRQn, GPIO_IRQ_PRIO);
-    NVIC_SetPriority(EXTI2_3_IRQn, GPIO_IRQ_PRIO);
-    NVIC_SetPriority(EXTI4_15_IRQn, GPIO_IRQ_PRIO);
+    /* enable the SYSCFG clock */
+    RCC->APB2ENR |= RCC_APB2ENR_SYSCFGEN;
```
Comment (haukepetersen) [inline cpu/stm32f0/periph/gpio.c]: same as line line 105, so you can remove this. Quote from the CPUs header file: "/* Old Bit definition maintained for legacy purpose */"....

File: cpu/stm32f0/periph/gpio.c [hunk 7]
```diff
@@ -24,315 +24,126 @@
...
+/**
+ * @brief   Allocate memory for one callback and argument per EXTI channel
+ */
+static exti_ctx_t exti_ctx[GPIO_ISR_CHAN_NUMOF];
 
-    port = gpio_port_map[dev];
-    pin = gpio_pin_map[dev];
+/**
+ * @brief   Extract the port base address from the given pin identifier
+ */
```
Comment (haukepetersen) [inline cpu/stm32f0/periph/gpio.c]: remove newline

File: cpu/stm32f0/periph/gpio.c [hunk 8]
```diff
@@ -24,315 +24,126 @@
...
+int gpio_init(gpio_t pin, gpio_dir_t dir, gpio_pp_t pullup)
+{
+    GPIO_TypeDef *port = _port(pin);
+    int pin_num = _pin_num(pin);
+
+    /* enable clock */
+    RCC->AHBENR |= (RCC_AHBENR_GPIOAEN << _port_num(pin));
+    /* configure pull register */
+    port->PUPDR &= ~(3 << (2 * pin_num));
+    port->PUPDR |= (pushpull << (2 * pin_num));
```
Comment (haukepetersen) [inline cpu/stm32f0/periph/gpio.c]: s/pushpull/pullup/

File: cpu/stm32f0/periph/gpio.c [hunk 9]
```diff
@@ -341,209 +152,77 @@ int gpio_init_int(gpio_t dev, gpio_pp_t pullup, gpio_flank_t flank, gpio_cb_t cb
     return 0;
 }
 
-void gpio_irq_enable(gpio_t dev)
+void gpio_init_af(gpio_t pin, gpio_af_t af)
```
Comment (haukepetersen) [inline cpu/stm32f0/periph/gpio.c]: the compiler says: 'gpio_af_t` undefined.

You need to add a typedef for this to `cpu/stm32f0/include/periph_cpu.h`. It needs to look like this:
```c


/**
 * @brief   Available MUX values for configuring a pin's alternate function
 */
typedef enum {
    GPIO_AF0 = 0,           /**< use alternate function 0 */
    GPIO_AF1,               /**< use alternate function 1 */
    GPIO_AF2,               /**< use alternate function 2 */
    GPIO_AF3,               /**< use alternate function 3 */
    GPIO_AF4,               /**< use alternate function 4 */
    GPIO_AF5,               /**< use alternate function 5 */
    GPIO_AF6,               /**< use alternate function 6 */
    GPIO_AF7,               /**< use alternate function 7 */
} gpio_af_t;
```

**Comment by haukepetersen** [comment]
Nice work! I can't test it yet, as the compiler still complains about some things, but they should be easy to fix...

**Comment by MohmadAyman** [comment]
thanks, i guess i've fixed all the logical mistakes, 
but how do i find the compilation errors and warnings ?, strider just tell that test failed with no further info.

**Comment by MohmadAyman** [comment]
hi @haukepetersen  i've fixed the errors in the code, but i do have a problem with squashing the commits, can you help me there ?

**Comment by PeterKietzmann** [comment]
Hi @MohmadAyman ! Did you already read our [git cheatsheet](https://github.com/RIOT-OS/RIOT/wiki/Git-cheatsheet) section "How to squash commits"?

**Comment by MohmadAyman** [comment]
thanks @PeterKietzmann , i found out that i had to fix the merge conflict first, now the static errors are all gone

**Comment by PeterKietzmann** [comment]
Cool! Could you please give that one commit a more speaking name? e.g. something like: "cpu/stm32f0: remodel gpio implementation"

**Comment by MohmadAyman** [comment]
i'm pretty puzzled here, i only changed the commit msg and then the strider tests succeeded ?!

**Comment by PeterKietzmann** [comment]
It's unrelated I guess :-) ! Currently Travis build is the choice. I'd say let's wait until someone tested and ACKed this PR (I won't before Monday) and then do the Travis build test (one needs to set a special label for starting the test)

**Comment by MohmadAyman** [comment]
Okay, but how do i set that label ?. i guess i can't add labels to the PR.

**Comment by PeterKietzmann** [comment]
That's correct. The reviewer will do it then

**Comment by MohmadAyman** [comment]
okay, so i fixed the static errors, and "BOARD=stm32f0discovery make" succeeded, is there anything else i can do to this PR?

**Comment by PeterKietzmann** [comment]
Did you test the functionality? Or don't you have the board available? If you have one it would surely be beneficial to check if things work on you side.

**Comment by MohmadAyman** [comment]
No, i don't have the board.

**Comment by MohmadAyman** [comment]
Needs testing.

**Comment by haukepetersen** [comment]
No building for the nucleo-f091:
```
/home/hauke/dev/riot/RIOT/cpu/stm32f0/periph/gpio.c: In function `gpio_set':
/home/hauke/dev/riot/RIOT/cpu/stm32f0/periph/gpio.c:174:15: error: `GPIO_TypeDef' has no member named `BSRRL'
     _port(pin)->BSRRL = (1 << _pin_num(pin));
               ^
/home/hauke/dev/riot/RIOT/cpu/stm32f0/periph/gpio.c: In function `gpio_clear':
/home/hauke/dev/riot/RIOT/cpu/stm32f0/periph/gpio.c:179:15: error: `GPIO_TypeDef' has no member named `BSRRH'
     _port(pin)->BSRRH = (1 << _pin_num(pin));
               ^
/home/hauke/dev/riot/RIOT/cpu/stm32f0/periph/gpio.c: In function `gpio_toggle':
/home/hauke/dev/riot/RIOT/cpu/stm32f0/periph/gpio.c:185:19: error: `GPIO_TypeDef' has no member named `BSRRH'
         _port(pin)->BSRRH = (1 << _pin_num(pin));
                   ^
/home/hauke/dev/riot/RIOT/cpu/stm32f0/periph/gpio.c:187:19: error: `GPIO_TypeDef' has no member named `BSRRL'
         _port(pin)->BSRRL = (1 << _pin_num(pin));
                   ^
/home/hauke/dev/riot/RIOT/cpu/stm32f0/periph/gpio.c: In function `gpio_write':
/home/hauke/dev/riot/RIOT/cpu/stm32f0/periph/gpio.c:194:19: error: `GPIO_TypeDef' has no member named `BSRRL'
         _port(pin)->BSRRL = (1 << _pin_num(pin));
                   ^
/home/hauke/dev/riot/RIOT/cpu/stm32f0/periph/gpio.c:196:19: error: `GPIO_TypeDef' has no member named `BSRRH'
         _port(pin)->BSRRH = (1 << _pin_num(pin));
```

**Comment by MohmadAyman** [comment]
Ok i see, i'll look into these errors.

**Comment by zhuoshuguo** [comment]
@MohmadAyman Ping. A kind reminder for rebase.

**Comment by MohmadAyman** [comment]
ok, I almost forgot about this PR.

**Comment by MohmadAyman** [comment]
@haukepetersen 
What just happened is that, ``` nucleo-f091 ``` uses stm32f091rc, and this CPU's GPIO_TypeDef registers are not the same as stm32f051x8 s',
I would change some of gpio.c functions to look like this
```
#ifdef __STM32F091xC_H
    _port(pin)->BSRR = (1 << _pin_num(pin));
#endif
#ifndef __STM32F051x8_H
    _port(pin)->BSRRL = (1 << _pin_num(pin));
#endif
```
When I add ```#define __STM32F091xC_H``` to nucleo-f091 board header, the GPIO works Fine!, But other things break -output below-, which would require me to edit some of nucleo-f091 code, do I proceed ?
```
/home/mohmmad/workspace/riot/gpio-riot/RIOT/boards/nucleo-f091/board.c: In function 'leds_init':
/home/mohmmad/workspace/riot/gpio-riot/RIOT/boards/nucleo-f091/board.c:46:5: error: 'RCC' undeclared (first use in this function)
     RCC->AHBENR |= RCC_AHBENR_GPIOAEN;
     ^
/home/mohmmad/workspace/riot/gpio-riot/RIOT/boards/nucleo-f091/board.c:46:5: note: each undeclared identifier is reported only once for each function it appears in
/home/mohmmad/workspace/riot/gpio-riot/RIOT/boards/nucleo-f091/board.c:46:20: error: 'RCC_AHBENR_GPIOAEN' undeclared (first use in this function)
     RCC->AHBENR |= RCC_AHBENR_GPIOAEN;
                    ^
In file included from /home/mohmmad/workspace/riot/gpio-riot/RIOT/boards/nucleo-f091/board.c:21:0:
/home/mohmmad/workspace/riot/gpio-riot/RIOT/boards/nucleo-f091/include/board.h:53:30: error: 'GPIOA' undeclared (first use in this function)
 #define LED_GREEN_PORT      (GPIOA)
```

**Comment by MohmadAyman** [comment]
I added some ``` #ifdef CPU_MODEL_STM32F091RC ``` and ``` #ifdef CPU_MODEL_STM32F051R8 ``` so the problem is gone.

**Comment by haukepetersen** [comment]
I would prefer if you just change one of the ST header files to make them use the same naming for the bit-set-reset registers. I don't see why we should make our code ugly, just because ST is not able to keep to a single naming scheme even inside a MCU family...

File: cpu/stm32f0/periph/gpio.c [hunk 10]
```diff
@@ -24,314 +24,107 @@
 #include "periph/gpio.h"
 #include "periph_conf.h"
 
+
+/**
+ * @brief   The STM32F3 has 16 EXTI channels
```
Comment (haukepetersen) [inline cpu/stm32f0/periph/gpio.c]: `STM32F0`

**Comment by haukepetersen** [comment]
Found an even cleaner solution to the problem: Just download the newest version of ST's Cube library (V1.4.0) and extract the header files for the `stm32f051x8.h` and `stm32f091xc.h`. Then update the existing headers in RIOT with these newer versions. Now you have the same GPIO registers for both... (but please to the update of vendor headers in it's own commit, so it's easier to track later on)

**Comment by MohmadAyman** [comment]
I was doing a rebase to squash the commits, and accidentally this happened, i don't know if this can be solved, so @haukepetersen , i can either open anther PR or  you can continue working on #4698  #4697 .

**Comment by PeterKietzmann** [comment]
@MohmadAyman something went wrong with your rebase. I'd propose to always work in branches as stated in [Contributing to RIOT](https://github.com/RIOT-OS/RIOT/wiki/Contributing-to-RIOT) (last section) or [Development procedures](https://github.com/RIOT-OS/RIOT/wiki/Development-procedures). We also have some quick help instructions for git in our wiki [Git cheatsheets](https://github.com/RIOT-OS/RIOT/wiki/Git-cheatsheet). Anyway, in general I would favour **your** PR as you started to work on it way earlier. However, if you won't find time to fix the PR soon I suggest to take @haukepetersen 's PR #4698 as he needs the driver update for further developments.

**Comment by MohmadAyman** [comment]
@PeterKietzmann  @haukepetersen , I did mange to get around the problem, this PR is ready now i guess.

**Comment by PeterKietzmann** [comment]
Thank

**Comment by PeterKietzmann** [comment]
Thanks Mohamad! @haukepetersen could you please have a look at the driver?

**Comment by haukepetersen** [comment]
I can live with this PR. Please rebase!

**Comment by haukepetersen** [comment]
The `test/periph_spi` is failing to build. Could you please fix this and squash your commits into some meaningful ones? I would suggest 3 commits, one for the CPU part and two for each changed board, using commit messages like `cpu/stm32f0: optimized GPIO driver implementation` or `boards/stm32f0discovery: adapted to GPIO driver changes` or something like that.

**Comment by haukepetersen** [comment]
@MohmadAyman: ping!

**Comment by MohmadAyman** [comment]
I don't understand, why would the nrf51 break when I'm only editing the stm32f0.

**Comment by haukepetersen** [comment]
could you rebase once more?!

About the Travis fail: I have no idea where this is coming from, I had similar trouble earlier today, locally on my computers everything is compiling just fine...

**Comment by MohmadAyman** [comment]
It worked.

**Comment by haukepetersen** [comment]
But the code does not. When running `tests/periph_gpio` on the `nucleo-f091`, it get hardfaults whenever calling `init_in` or `init_out`...

File: cpu/stm32f0/include/periph_cpu.h [hunk 11]
```diff
@@ -20,12 +20,76 @@
...
+
+/**
+ * @brief   Definition of a fitting UNDEF value
+ */
+#define GPIO_UNDEF          (0xffffffff)
+
+/**
+ * @brief   Define a CPU specific GPIO pin generator macro
+ */
+#define GPIO(x, y)          ((GPIOA_BASE + (x << 10)) | y)
```
Comment (haukepetersen) [inline cpu/stm32f0/include/periph_cpu.h]: This needs to be `GPIO_PIN(x, y)`. This is what cause the hardfaults...

**Comment by haukepetersen** [comment]
also external interrupts to not work on the `nucleo-f091`.

File: cpu/stm32f0/periph/gpio.c [hunk 12]
```diff
@@ -24,309 +24,96 @@
 #include "periph/gpio.h"
 #include "periph_conf.h"
 
+/**
+ * @brief   Allocate memory for one callback and argument per EXTI channel
+ */
 static gpio_isr_ctx_t gpio_config[GPIO_NUMOF];
```
Comment (haukepetersen) [inline cpu/stm32f0/periph/gpio.c]: `GPIO_NUMOF` is absolutely wrong here, this is a legacy value that defaults to `0` -> hence leading to not working external interrupts...

Change the top of this file to
```c
/**
 * @brief   The STM32F0 family has 16 external interrupt lines
 */
#define EXTI_NUMOF          (16U)

/**
 * @brief   Allocate memory for one callback and argument per EXTI channel
 */
static gpio_isr_ctx_t gpio_config[EXTI_NUMOF];
```
and see below on how to change the isr

File: cpu/stm32f0/periph/gpio.c [hunk 13]
```diff
@@ -336,209 +123,77 @@ int gpio_init_int(gpio_t dev, gpio_pp_t pullup, gpio_flank_t flank, gpio_cb_t cb
...
-    if (EXTI->PR & EXTI_PR_PR14) {
-        EXTI->PR |= EXTI_PR_PR14;        /* clear status bit by writing a 1 to it */
-        gpio_config[GPIO_IRQ_14].cb(gpio_config[GPIO_IRQ_14].arg);
-    }
-#endif
-#if GPIO_IRQ_15 >= 0
-    if (EXTI->PR & EXTI_PR_PR15) {
-        EXTI->PR |= EXTI_PR_PR15;        /* clear status bit by writing a 1 to it */
-        gpio_config[GPIO_IRQ_15].cb(gpio_config[GPIO_IRQ_15].arg);
+    for (int i = 0; i < GPIO_NUMOF; i++) {
```
Comment (haukepetersen) [inline cpu/stm32f0/periph/gpio.c]: change this to
```c
void isr_exti(void)
{
    for (size_t i = 0; i < EXTI_NUMOF; i++) {
        if (EXTI->PR & (1 << i)) {
            EXTI->PR = (1 << i);        /* clear by writing a 1 */
            gpio_config[i].cb(gpio_config[i].arg);
        }
    }
    if (sched_context_switch_request) {
        thread_yield();
    }
}
```

File: cpu/stm32f0/periph/gpio.c [hunk 14]
```diff
@@ -24,309 +24,96 @@
...
-            GPIO_10_EXTI_CFG();
-            break;
-#endif
-#ifdef GPIO_11_EN
-        case GPIO_11:
-            GPIO_11_EXTI_CFG();
-            break;
-#endif
-    }
-    NVIC_EnableIRQ(gpio_irq_map[dev]);
```
Comment (haukepetersen) [inline cpu/stm32f0/periph/gpio.c]: remove this newline

**Comment by haukepetersen** [comment]
Would be nice if you could address the comments above ASAP, as I am kind of waiting on this PR...

**Comment by haukepetersen** [comment]
And just squash your fixes right in.

**Comment by haukepetersen** [comment]
thanks. Works as expected on the `nucleo-f091` -> ACK once Travis is green.

**Comment by haukepetersen** [comment]
all green, go!

**Comment by MohmadAyman** [comment]
Sorry for the so many mistakes, and for my inexperience, and thanks for the help @haukepetersen .

**Comment by gebart** [comment]
@MohmadAyman don't excuse yourself for being a beginner, we have all been there. I bet you have learned alot during the evolution of this PR. Congratulations on getting it merged! Good job :+1:

**Comment by haukepetersen** [comment]
Don't worry, learning by doing is the way to go! Just have fun doing it :-)

**Comment by MohmadAyman** [comment]
Thank you guys!,  you really made me want to continue working on RIOT!.

**Pull request merged**